### PR TITLE
ISPN-10685 Align filesystem layout of the file-stores

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-10685

### CODE REVIEW

**Comment by tristantarrant** [comment]
Failures are unrelated

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java [hunk 1]
```diff
@@ -190,6 +193,26 @@ public void start() {
...
+      return ctx.getCache().getName().replaceAll("[^a-zA-Z0-9-_\\.]", "_");
+   }
+
+   protected String getQualifiedDataLocation() {
+      if (configuration.dataLocation() == null) {
+         return Paths.get(ctx.getGlobalStateConfiguration().persistentLocation(), sanitizedCacheName(), "data").toString();
+      } else {
+         return Paths.get(configuration.dataLocation(), sanitizedCacheName(), "data").toString();
+      }
+   }
```
Comment (ryanemerson) [inline persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java]: As these methods are pretty much the same for all stores, we should probably add a generic static method to `PersistenceUtil.java` in core.

Something like:

```java
public static String getQualifiedLocation(Supplier<String> location, GlobalStateConfiguration globalCfg, String cacheName, String dataDir) {
  // sanitize and return location
}
````

File: core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java [hunk 2]
```diff
@@ -74,4 +75,9 @@
     * Returns the persistence marshaller which should be used to marshall/unmarshall all stored bytes.
     */
    PersistenceMarshaller getPersistenceMarshaller();
+
+   /**
+    * Returns the global state configuration so that file-based stores can use default locations for storage
+    */
+   GlobalStateConfiguration getGlobalStateConfiguration();
```
Comment (ryanemerson) [inline core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java]: Maybe we should return the `GlobalConfiguration` here to allow greater flexibility? Some of the store implementations, jdbc for sure, have to retrieve the global config via the cache -> CacheManager currently.

Comment (tristantarrant) [reply core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java]: Done

File: persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java (hunk 1 above)
Comment (tristantarrant) [reply persistence/soft-index/src/main/java/org/infinispan/persistence/sifs/SoftIndexFileStore.java]: Done

File: core/src/main/java/org/infinispan/persistence/PersistenceUtil.java [hunk 3]
```diff
@@ -188,4 +190,30 @@ public static int count(SegmentedAdvancedLoadWriteStore<?, ?> salws, IntSet segm
...
+      return cacheName.replaceAll("[^a-zA-Z0-9-_\\.]", "_");
+   }
+
+   public static Path getQualifiedLocation(GlobalConfiguration globalConfiguration, String location, String cacheName, String qualifier) {
+      Path persistentLocation = Paths.get(globalConfiguration.globalState().persistentLocation());
+      if (location == null) {
+         return persistentLocation.resolve(Paths.get(sanitizedCacheName(cacheName), qualifier));
+      } else {
+         Path path = Paths.get(location);
+         if (path.isAbsolute()) {
```
Comment (anistor) [inline core/src/main/java/org/infinispan/persistence/PersistenceUtil.java]: I think the point of an absolute path is exactly that, to be able to get out of the 'cage'. Can't we trust users knowing what they do?

Comment (tristantarrant) [reply core/src/main/java/org/infinispan/persistence/PersistenceUtil.java]: But in managed environments you don't want to allow users creating caches wherever

Comment (tristantarrant) [reply core/src/main/java/org/infinispan/persistence/PersistenceUtil.java]: For flexibility, I can add a "strict persistent path" boolean attribute to global state which defaults to true, but which can be turned off.

Comment (ryanemerson) [reply core/src/main/java/org/infinispan/persistence/PersistenceUtil.java]: I'm more against having an additional switch than the restriction tbh, so if we're going todo it lets make it mandatory.

**Comment by ryanemerson** [comment]
Merging, thanks @tristantarrant. Let's go with the strict behaviour and if we have any complaints from users etc about usability, we can always relax this restriction in the future.

**Pull request merged**